Release 1.0.0

pRottinghuis opened the pull request:
Release 1.0.0
Initial stable release.

### Added
- Initial stable release.
- Check model file size before creating augmentation. Prevents creating invalid empty augmentation when large models are open.
- Documentation files added for commands and tool in the help menu.
- `scholar removeUser` command. Will delete the user and all related data from ChimeraX.
- Tool will prompt to delete an invalid user if they are selected from existing users.

### Changed
- Adjusted logger message statuses.
- Make `filePath` argument in `scholar saveAugSession` an `OpenFileNameArg` which allows passing `browse` to open file explorer.
- `scholar augmentation` makes a sub call to `scholar saveAugSession` to capture the current session when the augmentation is first created.
- `scholar cleanLocal` updated to have optional `username` field. Not passing any username will clean files for all local users.
- Clean local in the context menu on the tool will target the logged in user. If the tool is in the login screen, all users will be targeted and a warning will appear asking for a confirmation to clean files across all users.
- Hide `scholar login` command from the log to prevent API token from being in the log. Related info/warning messages also no longer display a token. Adds security for bug reporting with included log.

### Fixed
- Raise `NonChimeraXError` for any Schol-AR server issues.
- More detailed error logging for network calls.
- If an augmentation file upload fails, log now describes which target field failed.
- Log suggests actions for complying with file size and upload requirements.
- Popup warns user about potential dangers when trying to update target image.

pRottinghuis committed:
[rebase] Rebase develop onto main.

pRottinghuis committed:
[docs] Full project code documentation

pRottinghuis committed:
[#16012] Missing pre-augmentation creation file size check

Checks the file size of the current model befure creating an augmentation.
Will prevent the augmentation from being created and will alert the user
in the log should the file be too large to upload.

pRottinghuis committed:
[change] Convert log messages to warning

Changes warning log messages to be logged as warnings. Adjust no session found
to include 'yet'.

pRottinghuis committed:
[docs] Doc string store_qr_image method

Update doc string for store_qr_image in tool.py.

pRottinghuis committed:
[#16011] 500 Level Schol-AR network response

500 level network responses are indicated as NonChimeraX errors.
Added test coverage for response status levels.

pRottinghuis committed:
[change] Command saveAugSession file browse

Command saveAugSession file_path argument now is a OpenFileNameArg which allows browse.

pRottinghuis committed:
[16035] Schol-AR no session file found

When new augmentations are created there is automatically a new session created
for them. This means that work for an augmentation won't be lost between an augmentation
being created and the application closing without a session save. Also reduces
confusion in logging messages.

pRottinghuis committed:
[tests] Remove 200 status tests

Remove 200 status test from api_request tests.

pRottinghuis committed:
[refactor] Network request signature

Remove param from network call function that should
be implemented with try:except.

pRottinghuis committed:
[change] Log info and warnings

More specific log messaging. Suggestions are made
to the user in situations where user aciton is required.
Warning added to uploading a new target image. Update
tests to match new logging.

pRottinghuis committed:
[refactor] ScARFileManager lists usernames

Moves code to list usernames out of the tool and into the
file manager.

pRottinghuis committed:
[change] Command 'scholar cleanLocal' on all users

The 'scholar cleanLocal' command considers the username argument
optional. If no username argument is provided clean local will
remove extra data for all users on the device. In the GUI doing
clean local on the login page will act as a call with no username
argument, and doing cleann local on any page after the login
will target the logged in user only.

pRottinghuis committed:
[docs] Tool and command bundle documentation

HTML tool and command documentation

Addition of HTML files and supplementary
images and css for tool and command documentation.

pRottinghuis committed:
[change] Hide 'scholar login' in log

This is a safety feature to prevent api tokens from being
stored in the log. This will prevent api tokenns from getting
listed in issues posted to trac. The commannd is marked as
having it's own logger and does nothing with it.

pRottinghuis committed:
[feature] Command 'scholar removeUser'

The 'scholar removeUser' command allows for the removal of a
username api token pairing. Additional support for cases where
accounts are deleted from Schol-AR or if for some reason
api tokens were to ever change. Additionally, a user could remove
logins from their machine and all associated files. The
'scholar cleanLocal' command suggests removing users who's api
tokens are invalid.

pRottinghuis committed:
[docs] Markdown project documentation

Command and tool html files converted to markdown and added to projects README.md.
CHANGELOG.md added to reflect current version changes and changes from toolshed.

zjp commented on bundle_info.xml:
```diff
@@ -1,17 +1,42 @@
+<!--
+=== UCSF ChimeraX Copyright ===
+Copyright 2022 Regents of the University of California. All rights reserved.
```
nit: you should use 2024 as the beginning copyright year since that's when the code was written

pRottinghuis replied:
Updated.

zjp commented on src/cmd.py:
```diff
@@ -29,21 +58,23 @@ def login(session, username: str, api_token: Union[str, None] = None):
     ScARFileManager.update_users_info(username, user_token)
 
     ScARFileManager.update_user_projects(username)
+    session.logger.info("Succesfully logged into Schol-AR as: " + username)
```
This could be a status message, since when the UI populates with projects that makes it pretty clear if you've logged in or not -- but if you want to leave it in the log that's fine.

pRottinghuis replied:
Good point, I like it better as a status message too.

zjp commented on src/io.py:
```diff
@@ -37,12 +62,16 @@ class APIManager:
+
     @staticmethod
-    def try_api_request(request_fn, display_errors: bool, *args, **kwargs) -> Optional[dict]:
+    def try_api_request(request_fn, expecting_error=False, *args, **kwargs) -> Optional[dict]:
```
This expecting_error parameter is kind of a code smell. You should always expect the possibility that something will fail and handle it gracefully. If you know for sure that an invocation is going to fail somewhere, it's better to try...catch it when a comment in that location explaining why it's going to fail.

zjp replied:
try...catch is C++, sorry, I meant try...except.

pRottinghuis replied:
Makes sense. Refactored that param out.

zjp commented on src/io.py:
```diff
@@ -806,14 +1100,35 @@ def get_first_file(cls, search_dir: str) -> Optional[str]:
+        # Save the file to the temporary directory
+        file_name = f"scholar-file-check{file_extension}"
+        file_path = os.path.join(cls.TEMP_DIR, file_name)
+        run(session, f"save \"{file_path}\"", log=verbose)
```
nit: FYI: with f-strings you can use ' or " depending on what's convenient for you. If you use ' when you want to use " inside the string then you don't have to escape them.

pRottinghuis replied:
Noted.

zjp commented on src/io.py:
```diff
@@ -806,14 +1100,35 @@ def get_first_file(cls, search_dir: str) -> Optional[str]:
+        # Check if the file size is within the allowed limit
+        is_within_size_limit = cls.check_file_size(file_path)
+        # Delete the entire temporary directory
+        shutil.rmtree(cls.TEMP_DIR)
```
I'm not going to ask you to make a huge architectural change in this review, so you can close this comment without changes, but in the future if you find yourself making a class where every method is a classmethod, those methods could probably be static, and if every method is static and there's no data, that's not an object -- if you're making a class just to make a namespace consider a module of free functions instead. There's no need to be 'terminally object oriented'.

pRottinghuis replied:
That makes sense. I did end up making those changes.

zjp reviewed (commented):
LGTM overall. I had minor comments in the code. More broadly: squash your commits to 1 logical change per commit, consider shorter PRs. Big PRs make peoples' eyes glaze over, and make them want to type LGTM without a deep analysis.

zjp commented:
We'll let Greg review, too.

pRottinghuis committed:
[change] Login command status message

Login log message is now status message instead of info message.

pRottinghuis committed:
[refactor] Network request error handling refactor

Refactor try_api_request method. Removal of expecting_error param. New implementation for validate API which handles it's own error cases.

pRottinghuis committed:
[refactor] Api and file manager modules

Move all Schol-AR api and file manager related funcitonality into api_manager and sc_file_manager modules. General utilities for file actions used throughout the
project moved into utils module.

pRottinghuis committed:
[docs] Update license year

pRottinghuis commented:
@zjp thanks for taking a look. I made changes per your feedback.

pRottinghuis committed:
[docs] Toolshed link in README.md
